File: src/Web/Dart/example2.dart
// Abstract Product
abstract class Database {
  void connect();
  void query();
}

// Concrete Product
class Postgres implements Database {
  @override
  void connect() {
    print('Connecting to PostgreSQL');
  }

  @override
  void query() {
    print('Querying PostgreSQL');
  }
}

class MySQL implements Database {
  @override
  void connect() {
    print('Connecting to MySQL');
  }

  @override
  void query() {
    print('Querying MySQL');
  }
}

// Abstract Factory
abstract class DatabaseFactory {
  Database createDatabase();
}

// Concrete Factory
class PostgresFactory implements DatabaseFactory {
  @override
  Database createDatabase() {
    return Postgres();
  }
}

class MySQLFactory implements DatabaseFactory {
  @override
  Database createDatabase() {
    return MySQL();
  }
}

void useDatabase(DatabaseFactory factory) {
  var db = factory.createDatabase();
  db.connect();
  db.query();
}

// Usage
void main() {
  useDatabase(PostgresFactory());
  useDatabase(MySQLFactory());
}

File: src/Systems/Objective-C/example2.m
#import <Foundation/Foundation.h>

// Abstract Product
@protocol Database <NSObject>
- (void)connect;
- (void)query;
@end

// Concrete Product
@interface Postgres : NSObject <Database>
@end

@implementation Postgres
- (void)connect {
    NSLog(@"Connecting to PostgreSQL");
}
- (void)query {
    NSLog(@"Querying PostgreSQL");
}
@end

@interface MySQL : NSObject <Database>
@end

@implementation MySQL
- (void)connect {
    NSLog(@"Connecting to MySQL");
}
- (void)query {
    NSLog(@"Querying MySQL");
}
@end

// Abstract Factory
@protocol DatabaseFactory <NSObject>
- (id<Database>)createDatabase;
@end

// Concrete Factory
@interface PostgresFactory : NSObject <DatabaseFactory>
@end

@implementation PostgresFactory
- (id<Database>)createDatabase {
    return [Postgres new];
}
@end

@interface MySQLFactory : NSObject <DatabaseFactory>
@end

@implementation MySQLFactory
- (id<Database>)createDatabase {
    return [MySQL new];
}
@end

void useDatabase(id<DatabaseFactory> factory) {
    id<Database> db = [factory createDatabase];
    [db connect];
    [db query];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        useDatabase([PostgresFactory new]);
        useDatabase([MySQLFactory new]);
    }
    return 0;
}
